adoptopenjdk.groovy

```diff
@@ -41,41 +41,56 @@ class ListAdoptOpenJDK {
-            JSONObject r = new JSONObject()
-            r.put("name", "OpenJDK " + feature_version + " - " + jvm_impl)
-            JSONArray releases = new JSONArray()
-            int page = 0
-            boolean keepGoing = true
-            while (keepGoing) {
-                Page a = wc.getPage(API_URL + "/assets/feature_releases/" + feature_version + "/ga?vendor=eclipse&project=jdk&image_type=jdk&sort_method=DEFAULT&sort_order=DESC&page_size=20&page=" + (page++) + "&jvm_impl=" + openjdk_impl)
-                WebResponse ar = a.getWebResponse()
-                if (ar.getStatusCode() == 200) {
-                    JSONArray assets = JSONArray.fromObject(ar.getContentAsString())
-                    releases.addAll(assets.collect {
-                        [
-                                release_name: it.release_name,
-                                openjdk_impl: openjdk_impl,
-                                binaries    : it.binaries.collect {
-                                    [
-                                            architecture: it.architecture,
-                                            os          : it.os,
-                                            openjdk_impl: it.jvm_impl,
-                                            binary_link : it.package.link
-                                    ]
-                                }
-                        ]
-                    })
-                } else {
-                    keepGoing = false
-                }
-            }
-            r.put("releases", releases)
+            JSONObject r = getRelease(feature_version as String, jvm_impl)
             result.add(r)
         }
         return result
     }
+
+    private JSONObject getRelease(String feature_version, String jvm_impl){
+        String openjdk_impl = jvm_impl.toLowerCase()
+        JSONObject r = new JSONObject()
+        r.put("name", "OpenJDK " + feature_version + " - " + jvm_impl)
+        Map<String, JSONObject> releasesMap = new LinkedHashMap<>();
+        int page = 0
+        boolean keepGoing = true
+        while (keepGoing) {
+            Page a = wc.getPage(API_URL + "/assets/feature_releases/" + feature_version + "/ga?vendor=eclipse&project=jdk&image_type=jdk&sort_method=DEFAULT&sort_order=DESC&page_size=20&page=" + (page++) + "&jvm_impl=" + openjdk_impl)
+            WebResponse ar = a.getWebResponse()
+            if (ar.getStatusCode() == 200) {
+                JSONArray assets = JSONArray.fromObject(ar.getContentAsString())
+                assets.forEach { JSONObject asset -> releasesMap.merge(asset.release_name as String, asset, { v1, v2 ->
+                    if (v1.version_data?.adopt_build_number > v2.version_data?.adopt_build_number){
+                        v1.binaries.addAll(v2.binaries);
+                        return v1;
+                    } else {
+                        v2.binaries.addAll(v1.binaries);
+                        return v2;
+                    }
```

Comment:
Thanks very much for the quick response.  Can you confirm that I've understood the logic correctly?

I believe that if the v1.adopt_build_number is greater than the v2.adopt_build_number, then the (older) v2.binaries are included into the (newer) v1 binaries list and the (revised newer + older) v1 binaries list is returned.  Is my understanding correct that the newer list is preferred and the older values are added to the newer list?

```suggestion
                    if (v1.version_data?.adopt_build_number > v2.version_data?.adopt_build_number){
                        v1.binaries.addAll(v2.binaries);
                        return v1;
                    } else {
                        v2.binaries.addAll(v1.binaries);
                        return v2;
                    }
```

Reply:
> I believe that if the v1.adopt_build_number is greater than the v2.adopt_build_number, then the (older) v2.binaries are included into the (newer) v1 binaries list and the (revised newer + older) v1 binaries list is returned. Is my understanding correct that the newer list is preferred and the older values are added to the newer list?

yes, that was my intention.